libfreerdp/core/gateway/rdg.c

```diff
@@ -323,34 +359,313 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
+
+	Stream_SealLength(sWS);
+
+	status = tls_write_all(rdg->tlsOut, Stream_Buffer(sWS), Stream_Length(sWS));
+	Stream_Free(sWS, TRUE);
+
+	if (status < 0)
+		return FALSE;
+
+	return TRUE;
+}
+
+static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
+{
+	if (rdg->transferEncoding.isWebsocketTransport)
+		return rdg_write_websocket(rdg, sPacket);
+	else
+		return rdg_write_chunked(rdg, sPacket);
+}
+
+static int rdg_websocket_read(BIO* bio, BYTE* pBuffer, size_t size,
+                              rdg_http_websocket_context* encodingContext)
+{
+	int status;
+	int effectiveDataLen = 0;
+	assert(encodingContext != NULL);
+	while (TRUE)
+	{
+		switch (encodingContext->state)
+		{
+			case WebsocketStateOpcodeAndFin:
+			{
+				uint8_t buffer[1];
+				status = BIO_read(bio, (char*)buffer, 1);
+				if (status <= 0)
+					return (effectiveDataLen > 0 ? effectiveDataLen : status);
+
+				encodingContext->opcode = buffer[0];
+				if (((encodingContext->opcode & 0xf) != WebsocketContinuationOpcode) &&
+				    (encodingContext->opcode & 0xf) < 0x08)
+					encodingContext->fragmentOriginalOpcode = encodingContext->opcode;
+				encodingContext->state = WebsocketStateLengthAndMasking;
+			}
+			break;
+			case WebsocketStateLengthAndMasking:
+			{
+				uint8_t buffer[1];
+				status = BIO_read(bio, (char*)buffer, 1);
+				if (status <= 0)
+					return (effectiveDataLen > 0 ? effectiveDataLen : status);
+
+				encodingContext->masking = ((buffer[0] & WEBSOCKET_MASK_BIT) == WEBSOCKET_MASK_BIT);
+				encodingContext->lengthAndMaskPosition = 0;
+				encodingContext->payloadLength = 0;
+				uint8_t len = buffer[0] & 0x7f;
+				if (len < 126)
+				{
+					encodingContext->payloadLength = len;
+					encodingContext->state = (encodingContext->masking ? WebSocketStateMaskingKey
+					                                                   : WebSocketStatePayload);
+				}
+				else if (len == 126)
+					encodingContext->state = WebsocketStateShortLength;
+				else
+					encodingContext->state = WebsocketStateLongLength;
+			}
+			break;
+			case WebsocketStateShortLength:
+			case WebsocketStateLongLength:
+			{
+				uint8_t buffer[1];
+				uint8_t lenLength = (encodingContext->state == WebsocketStateShortLength ? 2 : 8);
+				while (encodingContext->lengthAndMaskPosition < lenLength)
+				{
+					status = BIO_read(bio, (char*)buffer, 1);
+					if (status <= 0)
+						return (effectiveDataLen > 0 ? effectiveDataLen : status);
+
+					encodingContext->payloadLength =
+					    (encodingContext->payloadLength) << 8 | buffer[0];
+					encodingContext->lengthAndMaskPosition += status;
+				}
+				encodingContext->state =
+				    (encodingContext->masking ? WebSocketStateMaskingKey : WebSocketStatePayload);
+			}
+			break;
+			case WebSocketStateMaskingKey:
+			{
+				WLog_WARN(
+				    TAG, "Websocket Server sends data with masking key. This is against RFC 6455.");
+				return -1;
+			}
+			break;
+			case WebSocketStatePayload:
+			{
+				uint8_t effectiveOpcode =
```

Comment:
I'd suggest this state to have its own treatment function as it's quite big right now